ethereum/core/src/main/java/org/hyperledger/besu/ethereum/proof/WorldStateProofProvider.java

```diff
@@ -92,4 +109,81 @@ public WorldStateProofProvider(final WorldStateStorage worldStateStorage) {
         b -> b,
         b -> b);
   }
+
+  public boolean isValidRangeProof(
+      final Bytes32 startKeyHash,
+      final Bytes32 endKeyHash,
+      final Bytes32 rootHash,
+      final List<Bytes> proofs,
+      final TreeMap<Bytes32, Bytes> keys) {
+
+    // check if it's monotonic increasing
+    if (!Ordering.natural().isOrdered(keys.keySet())) {
+      return false;
+    }
+
+    // when proof is empty we need to have all the keys to reconstruct the trie
+    if (proofs.isEmpty()) {
+      final MerklePatriciaTrie<Bytes, Bytes> trie =
+          new SimpleMerklePatriciaTrie<>(Function.identity());
+      // add the received keys in the trie
+      for (Map.Entry<Bytes32, Bytes> key : keys.entrySet()) {
+        trie.put(key.getKey(), key.getValue());
+      }
+
+      return rootHash.equals(trie.getRootHash());
+    }
+
+    // reconstruct a part of the trie with the proof
+    Map<Bytes32, Bytes> proofsEntries = Collections.synchronizedMap(new HashMap<>());
+    for (Bytes proof : proofs) {
+      proofsEntries.put(Hash.hash(proof), proof);
+    }
+
+    if (keys.isEmpty()) {
+      final MerklePatriciaTrie<Bytes, Bytes> trie =
+          new StoredMerklePatriciaTrie<>(
+              new InnerNodeDiscoveryManager<>(
+                  (location, hash) -> Optional.ofNullable(proofsEntries.get(hash)),
+                  Function.identity(),
+                  Function.identity(),
+                  startKeyHash,
+                  endKeyHash,
+                  false),
+              rootHash);
+      try {
+        // check is there is not missing element
+        trie.entriesFrom(startKeyHash, Integer.MAX_VALUE);
```

Comment:
does StoredEntriesCollector throw if it encounters a missing element? it must be deep.  Is there a way to make this more explicit with the functional version of `entriesFrom`?

Reply:
yes clearly too deep. I think it will be complicated to change the code. I update the comment let me know if it's better now